tensorflow/lite/micro/kernels/xtensa_hifimini/softmax.cc

```diff
@@ -1,3 +1,24 @@
+/*******************************************************************************
+  * Copyright (c) 2020 Cadence Design Systems, Inc.
+  *
+  * Permission is hereby granted, free of charge, to any person obtaining
+  * a copy of this software and associated documentation files (the
+  * "Software"), to use this Software with Cadence processor cores only and
+  * not with any other processors and platforms, subject to
+  * the following conditions:
+  *
+  * The above copyright notice and this permission notice shall be included
+  * in all copies or substantial portions of the Software.
+  *
+  * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+  * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+  * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+  * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+  * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+  * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+  * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+  ******************************************************************************/
+
 /* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
```

Comment:
General question about softmax - have you compared the lookup table based version with your optimized version?  I found that calculating exp(-x) is quite expensive in general, and since we have 8-bit inputs we can save a lot of cycles by e(-beta * scale * x) for all 256 possible inputs in a LUT populated in prepare().  If the LUT is faster, these optimizations would be awesome to have in prepare where the LUT is populated.  If the your optimized version is close in cycle count then it may not be worth the effort to convert to LUT based solution.

Reply:
Compared the cycles between Cadence optimization and LUT based solution.
Found that Cadence optimized cycles are better so I suggest to drop LUT based solution.